src/BlockTypesController.php

```diff
@@ -110,6 +110,12 @@ public function add_data_attributes( $content, $block ) {
 			if ( in_array( $key, $exclude_attributes, true ) ) {
 				continue;
 			}
+
+			// Attributes already persisted in HTML should be skipped.
+			if ( is_string( $value ) && is_int( strrpos( $content, $value ) ) ) {
+				continue;
+			}
+
```

Comment:
This is probably the trickiest part that I'm not sure about, basically, we don't append attributes that are already present on the HTML, this includes strings, but would also skip classNames, I'm not sure what this might omit as well, but I can't think of anything.

Reply:
I think we can just leave the duplication if it's harmless. It's less risky and could be removed in the future if we find a way to grab block attributes via JSON or some other way.